[WebGPU] iOS often reports 256MB max buffer size

mwyrzykowski opened the pull request:
#### 20eaa761a17cddfa7df1a0db999d84f9eff35bff
<pre>
[WebGPU] iOS often reports 256MB max buffer size
<a href="https://bugs.webkit.org/show_bug.cgi?id=297475">https://bugs.webkit.org/show_bug.cgi?id=297475</a>
<a href="https://rdar.apple.com/158431698">rdar://158431698</a>

Reviewed by Tadeu Zagallo.

As encountered on WebLLM and also reported on some other
AI / ML developers, WebGPU reporting a max buffer size of 1/10th
the device&apos;s true maximum buffer size is too limiting.

iPhone 16 Pro for instance can report a 4GB max buffer size from
Metal, so WebGPU would report 400MB.

Instead, allow up to 1 GB buffers on iOS as the jetsam limit is
still 1.5GB on iOS, so reporting any larger is going to result
in the web process getting instantly terminated due to memory pressure.

After this change, iPhone 16 Pro reports 1GB buffer size.

* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::maxBufferSize):

Canonical link: <a href="https://commits.webkit.org/298851@main">https://commits.webkit.org/298851@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1babd371e27e04f1925f3426a1ac27501dc75689

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116871 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122956 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67462 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/38e3be9f-afb1-4192-92bc-1b62281195d9) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37233 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45136 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88758 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9699171d-ee03-48ba-a5cd-efcb8a495dd0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119820 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29689 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104837 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69215 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28753 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66612 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99069 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23096 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126083 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43771 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32883 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97423 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44135 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101037 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97221 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24755 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42544 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20488 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43656 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49252 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43123 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46462 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44828 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f43b3b26d219329d18d35b3ae3128df0d8d03d97. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49472

mwyrzykowski commented:
This limitation was called out on https://medium.com/@marcelo.emmerich/webgpu-bugs-are-holding-back-the-browser-ai-revolution-27d5f8c1dfca and is really the only criticism feedback I observed for Safari's WebGPU implementation from the article.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5ce353f789c457a039ddb6d92cac11c4c62566f7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49472

mwyrzykowski committed:
[WebGPU] iOS often reports 256MB max buffer size
https://bugs.webkit.org/show_bug.cgi?id=297475
rdar://158431698

Reviewed by Tadeu Zagallo.

As encountered on WebLLM and also reported on some other
AI / ML developers, WebGPU reporting a max buffer size of 1/10th
the device's true maximum buffer size is too limiting.

iPhone 16 Pro for instance can report a 4GB max buffer size from
Metal, so WebGPU would report 400MB.

Instead, allow up to 1 GB buffers on iOS as the jetsam limit is
still 1.5GB on iOS, so reporting any larger is going to result
in the web process getting instantly terminated due to memory pressure.

After this change, iPhone 16 Pro reports 1GB buffer size.

* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::maxBufferSize):

Canonical link: https://commits.webkit.org/298851@main
